### ISSUE TITLE
Add support for *.dmg archives on OS X

### ISSUE THREAD

**Issue by mpscholten** [NixOS/nixpkgs#12969]
GUI-based applications on OS X are usually delivered as apple disk files (dmg). For example see [the OS X version of sublime text](https://www.sublimetext.com/3). It would be awesome to add support for these archives to nix's default `unpackPhase`. This way creating packages for OS X applications is way easier.

There is only one issue with adding support for dmg archives: dmg files are technically not archives. They are more like iso files. So we cannot unpack a dmg file, we have to it using `/usr/bin/hdiutil mount x.dmg`, then have to copy the content of the mounted disk and finally unmount the image again. This is not very clean, but this is the usual process of installing applications on OS X.

If nothing speaks against this feature I'd be happy to contribute this :)

**Comment by copumpkin** [comment]
I'm not sure it's worth adding to the default `unpackPhase` (mostly because we'll probably package apps differently from most packages anyway, and we're not building from source), but as a feature it's generally useful.

As an aside, I don't think we should use `hdiutil` or mounting for this. `libdmg-hfsplus` works fine for this sort of thing, is already packaged in nixpkgs, and doesn't require breaking out of the sandbox.

**Comment by mpscholten** [comment]
> but as a feature (under a separate name) it's generally useful.

Any suggestions how we could do it? Maybe like `fetchzip` (so e.g. `fetchdmg`)?

> As an aside, I don't think we should use hdiutil or mounting for this. libdmg-hfsplus works fine for this sort of thing, is already packaged in nixpkgs, and doesn't require breaking out of the sandbox.

Great suggestion, far better than mounting :)

**Comment by vcunat** [comment]
Maybe more like `unzip`. When you add it into build inputs, it adds the ability of `unpackPhase` to unpack zips. The (main) purpose of `fetchzip` is to fix nondeterminism in generated archives.

**Comment by mpscholten** [comment]
Ok thanks :) I'll provide a pr for this in the next days.

**Comment by mpscholten** [comment]
I've found out that `libdmg-hsfplus` is part of the `xpwn` package (`development/mobile/xpwn/default.nix`). So I've added a shell hook in the `xpwn` package:
```bash
unpackCmdHooks+=(_tryUnpackDmg)
_tryUnpackDmg() {
    if ! [[ "$curSrc" =~ \.dmg$ ]]; then return 1; fi
    dmg extract "$curSrc" image.hfs
    hfsplus image.hfs extractall .
}
```

Then I've added `setupHook = ./setup-hook.sh;` to `development/mobile/xpwn/default.nix`.

Now I've created a local `sublime3.nix` package to test it:
```
with import <nixpkgs> {};
let
    build = "3103";
in stdenv.mkDerivation {
    name = "sublimetext3-${build}";
    src = fetchurl {
        name = "sublime.dmg";
        url = "https://download.sublimetext.com/Sublime%20Text%20Build%20${build}.dmg";
        sha256 = "7eef51e1a344f238de2f69bb9f888e46ed9083133616249f31189db958f321b5";
    };
    phases = ["unpackPhase" "buildPhase" "installPhase"];
    buildInputs = [ xpwn ];
    installPhase = ''
        mkdir -p $out
        mkdir -p $out/Applications
        cp -R "Sublime Text.app" "$out/Applications"
    '';
    meta = {
    };
}
```

When running `nix-env -if sublime3.nix --option build-use-sandbox false` the `hfsplus image.hfs extractall .` step fails with `error: Cannot open volume.`. Any ideas why? [If you look at the xcode package where a dmg is also unpacked and installed](https://github.com/NixOS/nixpkgs/blob/master/pkgs/os-specific/darwin/xcode/default.nix#L25) you can see that the usage of `dmg` and `hfsplus` seems to be alright.

**Comment by matthewbauer** [comment]
@mpscholten Some dmg files are compressed with bzip2 or xz. The dmg tool will just spit out garbage if you give it a compressed dmg. File can usually detect which compression it is, then you need to unzip it like a single file before giving it to dmg.

**Comment by skrzyp** [comment]
You might also consider `libarchive` and its `bsdtar` util (very good one btw) which can detect and handle many types of archives.

**Comment by mpscholten** [comment]
Thanks for pointing this out.

Some thoughts on this: User experience is important for nix and nix adoption. User experience means that we do what the user expects. To do what the user expects the `_tryUnpackDmg` shell hook needs to be able to unpack any `.dmg` a user can unpack via the apple finder. This means adding a dependency to e.g. `libarchive`, but this dependency doesn't belong to `xpwn` which provides `libdmg-hsfplus`.

So either we don't add an dependency to e.g. `libarchive`, thus the UX suffers because the hook will not unpack all `.dmg`s out of the box, or we build a new package, e.g. called `undmg` which depends on `xpwn` and e.g. `libarchive` which then is able to unpack any `.dmg`.

I suggest creating a new shell script called `undmg` which tries to decompress and unpack the `dmg` file just like the apple finder, putting it on github and then creating a new nix package with a shell hook for this. Any thoughts on this? Or do you see a simpler or better way to get a great out of the box experience?

**Comment by matthewbauer** [comment]
Right now, unpackPhase is only using file extensions. Would it be possible to use magic numbers as well? I think file is available on every machine. In the end, it's going to be difficult to distinguish a .tar.gz file from a .dmg[.gz] file. Checking magic numbers and using file's mimetype output could be the way to go.

**Comment by mpscholten** [comment]
I guess this would work. This way we can be sure that we first need to unzip the file. Would you still go with the `undmg` approach (creating a new little package for this) I've outlined above, or do you have something other in mind?

**Comment by matthewbauer** [comment]
Yeah, we could fork the dmg part of xpwn. I'm kind of surprised there's nothing like that already. I was looking through to make sure libarchive didn't have anything. I suppose DMG files are almost always treated just as image files.

**Comment by matthewbauer** [comment]
Take a look at https://github.com/matthewbauer/undmg

I've got it working for gzip and bzip2 DMG files. Still working out some issues though.

**Comment by mpscholten** [comment]
@matthewbauer awesome! tried it out locally and seems to work perfect as expected, thanks :) I've just submitted [a pr which adds the package](https://github.com/NixOS/nixpkgs/pull/13636). Could you take a look over it? I guess we still need to wait until the issues you've mentioned are fixed until this can get merged.

**Comment by mpscholten** [comment]
This got implemented with #13636.
